r/src/bpcells-cpp/matrixIterators/H5DenseMatrixWriter.h

```diff
@@ -0,0 +1,143 @@
+    uint64_t max_capacity;
+    bool row_major;
+
+    static HighFive::DataSet createH5Matrix(
+        HighFive::File file,
+        std::string dataset_path,
+        uint64_t nrow,  // nrow of input matrix
+        uint64_t ncol,  // ncol of input matrix
+        bool row_major,
+        uint64_t chunk_size,
+        uint32_t gzip_level
+    ) {
+      HighFive::SilenceHDF5 s;
+      // Create a dataspace with initial shape and max shape
+      uint64_t nrow_h5 = nrow;
+      uint64_t ncol_h5 = ncol;
+      if (row_major) {
+          nrow_h5 = ncol;
+          ncol_h5 = nrow;
+      }
+      HighFive::DataSpace dataspace({nrow_h5, ncol_h5}, {HighFive::DataSpace::UNLIMITED, HighFive::DataSpace::UNLIMITED});
+
+      // Use chunking
+      HighFive::DataSetCreateProps props;
+      props.add(HighFive::Chunking(std::vector<hsize_t>{std::min(chunk_size, nrow_h5), std::min(chunk_size, ncol_h5)}));
+      if (gzip_level > 0) {
+          props.add(HighFive::Shuffle());
+          props.add(HighFive::Deflate(gzip_level));      
+      }
+
+      if (file.exist(dataset_path)) {
+          file.unlink(dataset_path);
+      }
+      // Create the dataset
+      return file.createDataSet<T>(dataset_path, dataspace, props);
+    }
+
+  public:
+    H5DenseMatrixWriter(
+        const HighFive::File &file,
+        std::string path,
+        uint64_t nrow,  // nrow of input matrix
+        uint64_t ncol,  // ncol of input matrix
+        bool row_major,
+        uint64_t max_capacity = 16384,
+        uint64_t chunk_size = 1024,
+        uint32_t gzip_level = 0
+    )
+        : dataset(createH5Matrix(file, path, nrow, ncol, row_major, chunk_size, gzip_level))
+        , max_capacity(max_capacity) 
+        , row_major(row_major) {}
+
+    void write(MatrixLoader<T> &mat_in, std::atomic<bool> *user_interrupt = NULL) override {
+
+        // Ensure that we write matrices sorted by row
+        OrderRows<T> mat((std::unique_ptr<MatrixLoader<T>>(&mat_in)));
+        // Don't delete our original matrix
+        mat.preserve_input_loader();
+
+        uint32_t col = 0;
+        uint64_t idx = 0; // Index of for col_ptr array
+        std::vector<T> zero_buf(mat_in.rows());
+        std::vector<T> val_buf = zero_buf;  // buffer for each column
+        while (mat.nextCol()) {
+            if (user_interrupt != NULL && *user_interrupt) return;
+            if (mat.currentCol() < col)
+                throw std::runtime_error("H5DenseMatrixWriter encountered out-of-order columns");
+
+            while (mat.load()) {
+                uint32_t i = 0;
+                while (i < mat.capacity()) {
+                    uint32_t capacity = std::min((uint32_t)max_capacity, mat.capacity() - i);
+                    for (uint32_t ii = 0; ii < capacity; ii++) {
+                        val_buf[*(mat.rowData() + i + ii)] = *(mat.valData() + i + ii);
+                    }
+                    idx += capacity;
```

Comment:
I think this section can actually be simplified a bit further. I don't think you need the inner loop or the `max_capacity` variable at all. In `StoredMatrixWriter`, an inner loop is used to minimize the buffer size for the downstream `NumWriter` objects, but in this case it's not needed since we are just staging everything in `val_buf` anyhow.

I also agree with Immanuel: you could also just replace `idx` with a `bool` that gets set to true if any data has been loaded from the column.